File: src/javascript/javascript-dsa/algorithm/tree-traversal.js
export default function BinaryTreeNode(val, left = null, right = null) {
  this.val = val;
  this.left = left;
  this.right = right;
}

/*
    Node -> Left -> Right
*/
export function preOrderTraversal(root) {
  if (root !== null) {
    console.log(root.val);
    preOrderTraversal(root.left);
    preOrderTraversal(root.right);
  }
}

/*
    Left -> Node -> Right
*/
export function inOrderTraversal(root) {
  if (root !== null) {
    inOrderTraversal(root.left);
    console.log(root.val);
    inOrderTraversal(root.right);
  }
}
/*
    Left -> Right -> Node
*/
export function postOrderTraversal(root) {
  if (root !== null) {
    postOrderTraversal(root.left);
    postOrderTraversal(root.right);
    console.log(root.val);
  }
}

File: src/python/algorithms/tree-traversal.py
def preorder(root, result):
    if not root:
        return
    result.append(root.val)
    preorder(root.left, result)
    preorder(root.right, result)


def preorderTraversal(root):
    res = []
    preorder(root, res)
    return res


def inorder(root, result):
    if not root:
        return None
    inorder(root.left, result)
    result.append(root.val)
    inorder(root.right, result)


def inorderTraversal(root):
    res = []
    inorder(root, res)
    return res


def postOrder(root, result):
    if not root:
        return
    postOrder(root.left, result)
    postOrder(root.right, result)
    result.append(root.val)


def postorderTraversal(root):
    res = []
    postOrder(root, res)
    return res
